alerts/machine-config-operator/MachineConfigControllerDrainError.md

```diff
@@ -0,0 +1,114 @@
+The machine-config-controller pod is the central point of management for incoming updates to 
+machines.
+
+For the commands below replace the $CONTROLLERPOD variable with the name of your own machine-config-controller pod name. 
+
+```console
+oc -n openshift-machine-config-operator logs $CONTROLLERPOD -c machine-config-controller
+```
+
+When a drain is initiated the following log will be recorded.
+
+```console
+1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: initiating drain
+```
+The controller will then log each and every pod drained from the node. 
+
+```console
+  1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Evicted pod <namespace>/<pod-name>
+```
+
+If the controller is unable to drain a pod after 1m30s then it will fire an error log.
+
+```console
+1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Waiting 1 minute then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
+```
+
+If the drain continues to fail then another failure log will fire.
+
+```console
+ 1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Drain has been failing for more than 10 minutes. Waiting 5 minutes then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
+```
+
+Finally if 1 hour has passed and the drain is still failing the following logs will be fired and the node will be marked degraded.
+
+```console
+1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
+```
+
+```console
+1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
+```
+
+The machine-controller logs will explicitly tell you what pods are failing to drain succesfully. Looking at the pods named in the logs is where you should start. 
+
+Common reasons a node cannot drain a pod successfully are:
+
+- The pod has a PodDisruptionBudget (PBD) set preventing the controller from deleting the pod.
+- The pod has storage attached and the kubelet is unable to unmount the storage succesfully.
+- The pod deletion may be prevented by a webhook if it's configured to target UPDATE operations or the webhook is failing to be called by the kube-apiserver.
+- The pod deletion could be stuck because the finalizers set in the pod are preventing it from terminating succesfully.
+
+## Mitigation 
+
+- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the old value after the upgrade completes.
+
+```console
+oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'
+```
```

Comment:
Should we make the following steps into a procedure?

Reply:
I'm hesitant on this one. Do you mean a more detailed procedure such as scaling to 0 then back up?

Reply:
I was thinking something like:

If a webhook is preventing the pod deletion:
1. Check the `openshift-kube-apiserver` logs 
 Failed calling webhook....
 
2. Check if it's a `mutating` or `validating` webhook.
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration

3. Backup and delete the webhook
$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
$ oc delete <webhook_type> <webhook_name>

Reply:
Oh you mean for the webhook? The PDB and webhook are separate points. I can add step numbers to the webhook if you think it's needed. But the PDB and webhook are not together that's why I have them marked with a - as they are different points.

Reply:
Actually, after looking at the [View File preview](https://github.com/openshift/runbooks/blob/7322f7892f7c57c9db08c0df54c0e7d904220107/alerts/machine-config-operator/MachineConfigControllerDrainError.md#mitigation), I think the Mitigation section is perfectly clear and easy to follow as you have it now. 
Just need some `+` characters to tuck the code blocks and text under the bulleted lines.

Reply:
@CourtCourt521 I see you used spaces to line up the steps and its code block. In docs we use a `+`. Does the spacing render properly?

. Check if it's a `mutating` or `validating` webhook.
+
 ```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```

Reply:
Hi,

No when I do it with my code it uses the + as a bullet point. It looks like this. 

- Check if it's a mutating or validating webhook.
+ 
```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```

Reply:
@mburke5678 ^